src/rpc/client.rs

```diff
@@ -5,43 +5,66 @@ use std::time::Duration;
 use std::collections::HashMap;
 use std::sync::{mpsc, Mutex, Arc};
 
+use super::handler::{DefaultHandler, Handler};
 use super::value::Value;
 
 use super::model;
 
 type Queue = Arc<Mutex<HashMap<u64, mpsc::Sender<Result<Value, Value>>>>>;
 
-pub struct Client<R: Read + Send + 'static, W: Write> {
+pub struct Client<R, W>
+    where R: Read + Send + 'static,
+          W: Write + Send + 'static
+{
     reader: Option<R>,
-    writer: W,
+    writer: Arc<Mutex<W>>,
     dispatch_guard: Option<JoinHandle<()>>,
     event_loop_started: bool,
     queue: Queue,
     msgid_counter: u64,
 }
 
-impl<R: Read + Send + 'static, W: Write> Client<R, W> {
+impl<R, W> Client<R, W>
+    where R: Read + Send + 'static,
+          W: Write + Send + 'static
+{
     pub fn take_dispatch_guard(&mut self) -> JoinHandle<()> {
         self.dispatch_guard.take().expect("Can only take join handle after running event loop")
     }
 
-    pub fn start_event_loop_cb<F: FnMut(&str, Vec<Value>) + Send + 'static>(&mut self, cb: F) {
-        self.dispatch_guard =
-            Some(Self::dispatch_thread(self.queue.clone(), self.reader.take().unwrap(), cb));
+    pub fn start_event_loop_handler<H>(&mut self, handler: Option<H>)
```

Comment:
@daa84 did you want to drop the `Option<H>` here and use `H` instead?  In order to do so, `DefaultHandler` must be exposed to the user, so they can pass along that value when they have no need of callbacks.